Sources/PitchSpeller/Data Structures/Graph.swift

```diff
@@ -109,69 +114,94 @@ public struct Graph <Value: Hashable>: Hashable {
-
+    private var nodeSet: Set<Node> = []
+    private var adjacencyMatrix: [NodePair: Edge] = [:]
+    
     // MARK: - Initializers
 
     /// Create a `Graph` with the given `adjacencyList`.
+    // FIXME: Consider renaming `incidenceList`.
     public init(_ adjacencyList: [Node: [Edge]] = [:]) {
-        self.adjacencyList = adjacencyList
+        self.nodeSet = Set(adjacencyList.keys)
+        for (node,edges) in adjacencyList {
+            for edge in edges {
+                self.adjacencyMatrix[NodePair(node,edge.destination)] = edge
+            }
+        }
+    }
+    
+    /// Create a `Graph` with the given `nodeSet` and `adjacencyMatrix`.
+    public init(_ nodes: [Node], _ adjacencyMatrix: [NodePair: Edge]) {
+        self.nodeSet = Set(nodes)
+        self.adjacencyMatrix = adjacencyMatrix
     }
 
     /// Create a `Graph` with the given `edges`.
     public init <S> (_ edges: S) where S: Sequence, S.Element == Edge {
         for edge in edges {
-            adjacencyList.safelyAppend(edge, toArrayWith: edge.source)
+            let source = edge.source
+            let destination = edge.destination
+            self.nodeSet.insert(source)
+            self.nodeSet.insert(destination)
+            self.adjacencyMatrix[NodePair(source,destination)] = edge
         }
     }
 
     /// Create a `Graph` with the given `nodes`.
     public init <S> (_ nodes: S) where S: Sequence, S.Element == Node {
-        for node in nodes {
-            adjacencyList[node] = []
-        }
+        self.nodeSet = Set(nodes)
     }
 
     // MARK: - Insance Methods
 
     /// Insert the given `node` into the graph.
     public mutating func insertNode(_ node: Node) {
-        if adjacencyList[node] == nil {
-            adjacencyList[node] = []
-        }
+        nodeSet.insert(node)
     }
 
     /// Add an edge from the given `source` to the given `destination` nodes, with the given
-    /// `value` (i.e., weight, or capacity). If the `value` of the edge is 0, the edge is removed.
-    public mutating func insertEdge(from source: Node, to destination: Node, value: Double) {
-        let edge = Edge(from: source, to: destination, value: value)
+    /// weight. If the `value` of the edge is 0, the edge is removed.
+    public mutating func insertEdge(from source: Node, to destination: Node, weight: Weight) {
+        let edge = Edge(from: source, to: destination, weight: weight)
         insertEdge(edge)
     }
 
     /// Add the given `edge` if it does not currently exist. Otherwise, replaces the edge with the
     /// equivalent `source` and `destination` nodes. If the `value` of the edge is 0, the edge is
     /// removed.
     public mutating func insertEdge(_ edge: Edge) {
-        removeEdge(from: edge.source, to: edge.destination)
-        if edge.value != 0 {
-            adjacencyList.safelyAppend(edge, toArrayWith: edge.source)
-        }
+        adjacencyMatrix[NodePair(edge.source, edge.destination)] = edge
```

Comment:
As discussed pseudo-offline, let's rip `NodePair` out into a generic `struct`:

```Swift
struct Pair <A,B> {
    let a: A
    let b: B
}
```

I might add the initializer:

```Swift
extension Pair {
    init(_ a: A, _ b: B) {
        self.a = a
        self.b = b
    }
}
```

Then, you can extend this to conform conditionally to a bunch of things:

```Swift
extension Pair: Equatable where A: Equatable, B: Equatable { }
extension Pair: Hashable where A: Hashable, B: Hashable { }
```

Then, we can remove the `NodePair` implementation, and replace the calls with `Pair<Node,Node>`.